lib/ring.ex

```diff
@@ -0,0 +1,651 @@
+    key(),
+    hash :: Hash.t(),
+    num :: non_neg_integer(),
+    back :: non_neg_integer(),
+    config :: Config.config()
+  ) :: {:ok, [Node.name()]} | {:error, atom()}
+  def do_find_stable_nodes(key, hash, num, back, config) do
+    Enum.reduce_while(0..back, {:ok, []}, fn
+      back, {:ok, []} ->
+        with {:ok, nodes} <- do_find_historical_nodes(key, hash, num, back, config) do
+          {:cont, {:ok, nodes}}
+        end
+
+      back, {:ok, acc} ->
+        with {:ok, nodes} <- do_find_historical_nodes(key, hash, num, back, config) do
+          acc =
+            nodes
+            |> Enum.reverse()
+            |> Enum.reduce(acc, fn node, acc ->
+              if node in acc do
+                acc
+              else
+                [node | acc]
+              end
+            end)
+          {:cont, {:ok, acc}}
+        else
+          error ->
+            {:halt, error}
+        end
+    end)
+  end
+
+  @spec do_gc(table :: :ets.tid(), generation :: Config.generation()) :: :ok
+  defp do_gc(table, generation) do
+    :ets.match_delete(table, {{generation, :_}, :_})
+    :ok
+  end
+
+  @spec find_next_highest_item(table :: :ets.tid, generation :: Config.generation(), size(), hash :: Hash.t()) :: Node.virtual()
+  defp find_next_highest_item(_table, _generation, 0, _hash) do
+    nil
+  end
+
+  defp find_next_highest_item(table, generation, _size, hash) do
+    key =
+      case :ets.next(table, {generation, hash}) do
+        {^generation, _hash} = key ->
+          key
+
+        _ ->
+          # Generation is exhausted, start back up at the top of this generation.
+          :ets.next(table, {generation, -1})
+      end
+
+    case :ets.lookup(table, key) do
+      [{{^generation, hash}, name}] ->
+        {hash, name}
+
+      _ ->
+        nil
+    end
+  end
+
+  @spec has_node_with_name?(nodes :: [Node.t()], name :: Node.name()) :: boolean()
+  defp has_node_with_name?(nodes, name) do
+    Enum.any?(nodes, &match?({^name, _}, &1))
+  end
+
+  @spec update_nodes(state :: t(), nodes :: [Node.t()]) :: t()
+  defp update_nodes(%__MODULE__{} = state, nodes) do
+    next_generation = state.generation + 1
+
+    # Generate items for the next generation
+    items =
+      nodes
+      |> Node.expand()
+      |> Enum.map(fn {hash, name} ->
+        {{next_generation, hash}, name}
+      end)
+
+    # Write the items into the table for this generation
+    :ets.insert(state.table, items)
+
+    # Add the new size to the sizes
+    sizes = [Enum.count(nodes) | state.sizes]
+
+    # Truncate sizes to fit into depth
+    sizes = Enum.take(sizes, state.depth)
+
+    # Update the configuration to atomically cut over to the new generation
+    config = {
+      state.table,
+      state.depth,
+      sizes,
+      next_generation,
+      state.overrides
+    }
+
+    Config.set(state.name, self(), config)
+
+    # Schedule the stale generation for cleanup
+    state = schedule_gc(state, next_generation - state.depth)
```

Comment:
doesn't this need to be 
```suggestion
    state = schedule_gc(state, next_generation - state.depth - 1)
```
without the `-1` wouldn't this be off by 1?
e.g.
```
state.depth = 1
next_generation = 10
result = 9
```
But you want to use `9` as your backup?

Reply:
A ring with depth 1 only keeps around 1 generation, the current generation.

Perhaps this is confusing, do you think depth should exclude the current generation so a ring that has no history is depth 0 instead of depth 1?